File: services/stock-rust-salvo/src/service.rs
use crate::{
    model::{Category, NewCategory, UpdateCategory},
    repository,
};
use anyhow::Result;
use uuid::Uuid;

pub struct Service {
    repository: repository::Repository,
}

impl Service {
    pub fn new(repository: repository::Repository) -> Self {
        Service {
            repository: repository,
        }
    }
    pub fn get_categories(&self) -> Result<Vec<Category>> {
        self.repository.get_categories()
    }
    pub fn create_category(&self, category: NewCategory) -> Result<Category> {
        self.repository.create_category(&category)
    }
    pub fn update_category(&self, id: Uuid, category: UpdateCategory) -> Result<Category> {
        self.repository.update_category(&id, &category)
    }
    pub fn get_category(&self, id: Uuid) -> Result<Category> {
        self.repository.get_category(&id)
    }
    pub fn delete_category(&self, id: Uuid) -> Result<()> {
        self.repository.delete_category(&id)
    }
}

File: services/stock-go/internal/service/service.go
package service

import (
	"context"

	"github.com/jdotw/warehouse/services/stock-go/internal/repository"
	"github.com/jdotw/warehouse/services/stock-go/pkg/model"
)

type Service interface {
	GetCategories(ctx context.Context) (*[]model.Category, error)
	CreateCategory(ctx context.Context, category *model.Category) (*model.Category, error)
	DeleteCategory(ctx context.Context, categoryID string) error
	GetCategory(ctx context.Context, categoryID string) (*model.Category, error)
	UpdateCategory(ctx context.Context, category *model.Category) (*model.Category, error)
}

type service struct {
	r repository.Repository
}

func NewService(r repository.Repository) (Service, error) {
	return &service{r: r}, nil
}

func (s service) GetCategories(ctx context.Context) (*[]model.Category, error) {
	c, err := s.r.GetCategories(ctx)
	return c, err
}
func (s service) CreateCategory(ctx context.Context, category *model.Category) (*model.Category, error) {
	c, err := s.r.CreateCategory(ctx, category)
	return c, err
}
func (s service) DeleteCategory(ctx context.Context, categoryID string) error {
	err := s.r.DeleteCategory(ctx, categoryID)
	return err
}
func (s service) GetCategory(ctx context.Context, categoryID string) (*model.Category, error) {
	c, err := s.r.GetCategory(ctx, categoryID)
	return c, err
}
func (s service) UpdateCategory(ctx context.Context, category *model.Category) (*model.Category, error) {
	c, err := s.r.UpdateCategory(ctx, category)
	return c, err
}
